### ISSUE TITLE
fieldaligment for generated protobuf structs

### ISSUE THREAD

**Issue by vtolstov** [golang/protobuf#1392]
**What version of protobuf and what language are you using?**
Version: latest master

i'm generate protobuf go code from 
```
message CreatePosIssueReq {
  google.protobuf.Value fields = 1 [json_name="fields"];
};
```
**What did you expect to see?**

properly aligned go struct to minimize memory usage

**What did you see instead?**
```
type CreatePosIssueReq struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Fields *structpb.Value `protobuf:"bytes,1,opt,name=fields,proto3" json:"fields,omitempty"`
}
```

fieldaligment suggest struct with 40 pointer bytes could be 16

**Comment by vtolstov** [comment]
i think that overhead for each message is about 34 bytes is very ugly

**Comment by vtolstov** [comment]
any comments from golang protobuf team ?

**Comment by puellanivis** [comment]
Looking into this, for `GOARCH=amd64`, each field is aligned to 8-byte boundaries. `unknownFields` being a slice is 24 bytes long, `sizeCache` being a `uint32` is 4-bytes long, and `state` being a pointer is 8-bytes. There is no way to reorder the fields in order to save any space over the 40 bytes taken up by the fields `state, sizeCache, unknownFields`. https://go.dev/play/p/8D-_La9WXkR

Porting this out of Go playground so I can get 32-bit `GOARCH=386`, both `state` and `sizeCache` are 4-bytes in length, and `unknownFields` is 12-bytes. This is 20-bytes taken up by the fields `state, sizeCache, unknownFields` which being aligned to 4-byte boundaries… still comes up with 20-bytes for all orderings of arguments.

Running `fieldalignment` locally, it does appear that I get a similar result: `48 pointer bytes could be 24`. And running with `-fix` to apply the recommended fix results in a reordering that still has the same total length: 48 bytes.

The problem being raised here by the tool is that the garbage collector has to scan through all of `state, sizeCache, unknownFields` before getting to `Fields` to check that pointer. _This does not mean the struct will take up less space if reordered._ By placing the pointer values first (`state, Fields, unknownFields`) we pack the number of bytes scanned by the garbage collector into a smaller area.

However, ordering the exported fields into the unexported struct entries is well beyond a reasonable ask for this code generator. If this sort of pointer-scan overhead in the garbage collection is found to be a bottleneck to performance, then it should be recommended that the user runs `fieldalignment -fix` on their generated output. Thus allowing a single authoritative tool to reorder these fields, rather than having the `protoc-gen-go` attempt to replicate the same sorting.

There simply isn’t much here that is reasonably actionable by the go protobuf team. Reordering the preamble fields won’t save any actual memory usage, and while ordering them `state, unknownFields, sizeCache` would reduce the pointer-scan overhead for empty messages, any field that constructs a pointer value (i.e. all values except simple scalars without the `optional` keyword) would cause the fieldalignment tool to complain about the ordering. At that point, run the tool with `-fix` if the bottleneck becomes a performance bottleneck.

TL;DR: There is no reasonable space saving to be had here. The savings indicated by `fieldalignment` is wholly due to garbage collection scanning, which is usually not a primary concern for performance. In such cases where that performance might matter, the user already has an automated tool to reorder the fields, putting any amount of work into replicating such behavior is therefore unwise.

**Comment by vtolstov** [comment]
Ok looks reasonable. Thank you.
